### PR TITLE
ml(svm): fix UB pointer overflow in solve_generic with extreme float …

### LINKED ISSUE #29110: Invalid Index (UB Pointer Overflow) in SVM `solve_generic` via Extreme Float Training Data
### System Information

Environment: Dev Container on Linux
OS: Ubuntu 24.04.4 LTS (Noble Numbat)
Kernel: 7.0.0-15-generic
Architecture: x86_64
Model: Intel(R) Xeon(R) CPU Max 9468
clang++: Ubuntu clang version 21.1.8
cmake: 3.28.3
ninja: 1.11.1
GNU make: 4.3
Python: 3.12.13
Git: 2.43.0
UBSAN_OPTIONS: halt_on_error=1:print_stacktrace=1
Sanitizers used: address, undefined

#29111  tries to fix this bug

## Summary

A UBSan pointer arithmetic overflow occurs in `cv::ml::SVMImpl::Solver::solve_generic` when training a `NU_SVC` SVM with training data containing near-`FLT_MAX` float values. The internal kernel computation produces `Inf` gradients, causing `select_working_set_nu_svm` to return a stale index of `-1` while still reporting "not converged". The caller then dereferences `samples.ptr<float>(-1)`, triggering undefined behavior.

## Version

```
$ git describe
4.13.0-150-gc7732e1043
```

## Description

The bug is a combination of two flaws in `modules/ml/src/svm.cpp`:

### Flaw 1 — `select_working_set_nu_svm` can return index `-1` with return code `0`

The function initializes all four candidate indices to `-1` and updates them only when a gradient comparison succeeds. When all gradient values `G[]` are `Inf` (due to float overflow in the kernel), no comparison succeeds and all indices remain at `-1`. The convergence check `MAX(Gmax1+Gmax2, Gmax3+Gmax4) < eps` evaluates to `Inf < eps = false`, so the function returns `0` ("not converged, keep iterating") with uninitialized indices:

```cpp
// svm.cpp:893 — select_working_set_nu_svm
int Gmax1_idx = -1;   // never updated when G[] == Inf
int Gmax2_idx = -1;
int Gmax3_idx = -1;
int Gmax4_idx = -1;
...
// convergence check fails (Inf < eps is false), falls through to:
out_i = Gmax1_idx;   // = -1
out_j = Gmax2_idx;   // = -1
return 0;            // tells caller "valid indices found" — but they are not
```

### Flaw 2 — `solve_generic` passes the index to `get_row` without validation

After calling the working set selector, `solve_generic` unconditionally calls `get_row(i, ...)` with no bounds check on `i`:

```cpp
// svm.cpp:700-703 — solve_generic
if( (this->*select_working_set_func)( i, j ) != 0 || iter++ >= max_iter )
    break;
Q_i = get_row( i, &buf[0][0] );   // i = -1, no guard
Q_j = get_row( j, &buf[1][0] );
```

`get_row(-1)` calls `get_row_base(-1)` which calls `samples.ptr<float>(-1)`:

```cpp
// svm.cpp:564 — get_row_base
kernel->calc( sample_count, var_count, samples.ptr<float>(),
              samples.ptr<float>(i1),   // i1 = -1 → UB pointer overflow
              lru_cache_data.ptr<Qfloat>(kr.idx) );
```

### How extreme floats trigger Flaw 1

The LINEAR kernel computes dot products of the training vectors (`~3.4e38 * ~3.4e38`), which overflows to `Inf` (svm.cpp:177–181). These `Inf` values propagate into the gradient vector `G[]` during the initialization loop of `solve_generic` (svm.cpp:669–677). The Release-build overflow guard that would catch this is inside `#ifdef _DEBUG` and therefore inactive:

```cpp
// svm.cpp:686-694 — inactive in Release build
#ifdef _DEBUG
    for( i = 0; i < alpha_count; i++ )
    {
        if( fabs(G[i]) > 1e+300 )   // would catch Inf gradients
            return false;
        ...
    }
#endif
```

## PoC Code

```cpp
// Minimal PoC — no fuzzer infrastructure required
// Trigger: NU_SVC SVM trained on 6 samples with near-FLT_MAX feature values
#include <opencv4/opencv2/core.hpp>
#include <opencv4/opencv2/ml.hpp>

int main() {
    float fdata[] = {
        -3.40282347e+38f,  0.0f,
         3.08375703e+38f,  0.0f,
         3.38958311e+38f,  0.0f,
        -3.40053886e+38f,  0.0f,
        -5.19229686e+33f,  0.0f,
        -5.17201445e+33f,  0.0f,
    };
    int labels[] = {1, 0, 0, 0, 0, 0};

    cv::Mat features(6, 2, CV_32F, fdata);
    cv::Mat responses(6, 1, CV_32S, labels);

    cv::Ptr<cv::ml::SVM> svm = cv::ml::SVM::create();
    svm->setType(101);   // NU_SVC
    svm->setKernel(0);   // LINEAR
    svm->setNu(0.010317);
    svm->setTermCriteria(cv::TermCriteria(
        cv::TermCriteria::MAX_ITER | cv::TermCriteria::EPS, 11, 0.099998));

    svm->train(features, cv::ml::ROW_SAMPLE, responses);
    return 0;
}
```

## Reproduction Steps

```bash
OUTPUT=<path-to-opencv-sanitizer-build>

clang++ -g -O0 -fstandalone-debug -fno-omit-frame-pointer \
    -fsanitize=address,undefined \
    -I$OUTPUT/include -I$OUTPUT/include/opencv4 \
    poc.cpp -o poc \
    -Wl,--start-group \
        $OUTPUT/lib/lib*.a \
        $OUTPUT/3rdparty/lib/lib*.a \
    -Wl,--end-group \
    -lz -ldl -lpthread -lm

export UBSAN_OPTIONS=halt_on_error=1:print_stacktrace=1
./poc
```

## Stack Trace

```
modules/core/include/opencv2/core/mat.inl.hpp:734:24:
    runtime error: addition of unsigned offset to 0x... overflowed to 0x...
    #0  float* cv::Mat::ptr<float>(int)
            mat.inl.hpp:734
    #1  cv::ml::SVMImpl::Solver::get_row_base(int, bool*)
            svm.cpp:564
    #2  cv::ml::SVMImpl::Solver::get_row(int, float*)
            svm.cpp:632
    #3  cv::ml::SVMImpl::Solver::solve_generic(SolutionInfo&)
            svm.cpp:703
    #4  cv::ml::SVMImpl::Solver::solve_nu_svc(...)
            svm.cpp:1072
    #5  cv::ml::SVMImpl::do_train(cv::Mat const&, cv::Mat const&)
            svm.cpp:1505
    #6  cv::ml::SVMImpl::train(cv::Ptr<cv::ml::TrainData> const&, int)
            svm.cpp:1632
    #7  cv::ml::StatModel::train(cv::_InputArray const&, int, cv::_InputArray const&)
            inner_functions.cpp:74
    #8  main  poc.cpp:31

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior mat.inl.hpp:734:24
```

## Suggested Fix

Two complementary fixes are needed:

**Fix 1 (primary)** — Guard against stale `-1` indices in `solve_generic` before calling `get_row`:

```diff
--- a/modules/ml/src/svm.cpp
+++ b/modules/ml/src/svm.cpp
@@ -700,6 +700,8 @@ bool cv::ml::SVMImpl::Solver::solve_generic( SolutionInfo& si )
         if( (this->*select_working_set_func)( i, j ) != 0 || iter++ >= max_iter )
             break;
+        if( i < 0 || j < 0 )
+            break;
         Q_i = get_row( i, &buf[0][0] );
         Q_j = get_row( j, &buf[1][0] );
```

**Fix 2 (defense-in-depth)** — Move the gradient overflow early-exit out of `#ifdef _DEBUG` so it is active in Release builds:

```diff
--- a/modules/ml/src/svm.cpp
+++ b/modules/ml/src/svm.cpp
@@ -685,7 +685,7 @@ bool cv::ml::SVMImpl::Solver::solve_generic( SolutionInfo& si )
-        #ifdef _DEBUG
         for( i = 0; i < alpha_count; i++ )
         {
             if( fabs(G[i]) > 1e+300 )
                 return false;
             if( fabs(alpha[i]) > 1e16 )
                 return false;
         }
-        #endif
```

---
> Signed-off-by: FuzzAnything fuzzanything@gmail.com

### Issue submission checklist

- [x] I report the issue, it's not a question
- [x] I checked the problem with documentation, FAQ, open issues, forum.opencv.org, Stack Overflow, etc and have not found any solution
- [x] I updated to the latest OpenCV version and the issue is still there
- [x] There is reproducer code and related data files (videos, images, onnx, etc)

### PR DESCRIPTION
**Pull request by pranayr710**
…data

Fix #29110: When training NU_SVC SVM with near-FLT_MAX float values, the LINEAR kernel overflows to Inf. This causes select_working_set_nu_svm to return stale -1 indices, leading to undefined behavior in get_row(-1).

Two fixes:
1. Add bounds check for invalid indices in solve_generic before get_row
2. Move gradient overflow early-exit out of #ifdef _DEBUG so it activates in Release builds

Includes regression test.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

# Changes Made

### Fix 1 (lines 689–696) — Moved Gradient Overflow Early-Exit Out of `#ifdef _DEBUG`

```diff
-       #ifdef _DEBUG
                for( i = 0; i < alpha_count; i++ )
                {
                    if( fabs(G[i]) > 1e+300 )
                        return false;
                    if( fabs(alpha[i]) > 1e16 )
                        return false;
                }
-       #endif
```

This ensures the overflow check runs in all builds (not just debug), catching `Inf` gradients before they poison the working set selection.

---

### Fix 2 (lines 701–702) — Added Bounds Check for Invalid Indices

```diff
if( (this->*select_working_set_func)( i, j ) != 0 || iter++ >= max_iter )
                    break;
+
+               if( i < 0 || j < 0 )
+                   break;
```

This prevents:

```cpp
get_row(-1) → samples.ptr<float>(-1)
```

undefined behavior if `select_working_set_nu_svm` returns stale `-1` indices.

---

## 2. `test_svmtrainauto.cpp` — Regression Test

Added:

```cpp
ML_SVM.extreme_float_no_crash_29110
```

test (lines 164–189) using the exact PoC data from the bug report.

The test verifies that SVM training completes without crashing.

---

# Verification

- ✅ `git diff --check` — no whitespace issues
- ✅ Changes are minimal:
  - 2 files modified
  - `+30` lines
  - `-2` lines
- ✅ Committed on branch:
  ```bash
  fix/svm-ub-overflow-29110
  ```
- ✅ Branch based on:
  ```bash
  upstream/4.x
  ```
- ✅ Pushed to fork:
  ```bash
  pranayr710/opencv
  ```

### CODE REVIEW

**Comment by pranayr710** [comment]
The `precommit_docs` failure is unrelated to this PR. The build log shows:

[javadoc] javadoc: error - Error fetching URL: https://docs.oracle.com/en/java/javase/11/docs/api/ BUILD FAILED



This is a network connectivity issue on the CI buildbot — it cannot reach Oracle's 
Java API docs URL. This PR only modifies `modules/ml/src/svm.cpp` (internal logic) 
and `modules/ml/test/test_svmtrainauto.cpp` (test). No documentation was changed.

The **Linux x64 Debug** build (#113306) passed successfully ✅.

**Comment by pranayr710** [comment]
> The `precommit_docs` failure is unrelated to this PR. The build log shows:
> 
> [javadoc] javadoc: error - Error fetching URL: https://docs.oracle.com/en/java/javase/11/docs/api/ BUILD FAILED
> 
> This is a network connectivity issue on the CI buildbot — it cannot reach Oracle's Java API docs URL. This PR only modifies `modules/ml/src/svm.cpp` (internal logic) and `modules/ml/test/test_svmtrainauto.cpp` (test). No documentation was changed.
> 
> The **Linux x64 Debug** build (#113306) passed successfully ✅.

**Comment by asmorkalov** [comment]
Duplicates https://github.com/opencv/opencv/pull/29111
